update-engine/src/context.rs

```diff
@@ -24,30 +28,112 @@ use crate::{events::StepProgress, StepSpec};
 /// likely that it is dropped at the same time the future completes.
 #[derive(Debug)]
 pub struct StepContext<S: StepSpec> {
-    progress_sender: mpsc::Sender<StepProgress<S>>,
+    log: slog::Logger,
+    payload_sender: mpsc::Sender<StepContextPayload<S>>,
     token: StepHandleToken<S>,
 }
 
 impl<S: StepSpec> StepContext<S> {
-    pub(crate) fn new(progress_sender: mpsc::Sender<StepProgress<S>>) -> Self {
-        Self { progress_sender, token: StepHandleToken::new() }
+    pub(crate) fn new(
+        log: &slog::Logger,
+        payload_sender: mpsc::Sender<StepContextPayload<S>>,
+    ) -> Self {
+        Self { log: log.clone(), payload_sender, token: StepHandleToken::new() }
     }
 
     /// Sends a progress update to the update engine.
     #[inline]
     pub async fn send_progress(&self, progress: StepProgress<S>) {
-        self.progress_sender
-            .send(progress)
+        self.payload_sender
+            .send(StepContextPayload::Progress(progress))
             .await
             .expect("our code always keeps the receiver open")
     }
 
+    /// Creates a nested execution engine.
+    ///
+    /// An individual step can generate other steps: these steps are treated as
+    /// *nested*, and carry their own progress.
+    pub async fn with_nested_engine<'a, 'this, F, S2>(
+        &'this self,
+        engine_fn: F,
+    ) -> Result<CompletionContext<S2>, S2::Error>
+    where
+        'this: 'a,
+        F: FnOnce(&mut UpdateEngine<'a, S2>) -> Result<(), S2::Error> + Send,
+        S2: StepSpec,
+    {
+        let (sender, mut receiver) = mpsc::channel(128);
+        let mut engine = UpdateEngine::new(&self.log, sender);
+        // Create the engine's steps.
+        (engine_fn)(&mut engine)?;
+
+        // Now run the engine.
+        let engine = engine.execute();
+        tokio::pin!(engine);
+
+        let mut result = None;
+        let mut events_done = false;
+
+        loop {
+            tokio::select! {
+                ret = &mut engine, if result.is_none() => {
+                    match ret {
+                        Ok(cx) => {
+                            result = Some(Ok(cx));
+                        }
+                        Err(ExecutionError::EventSendError(_)) => {
+                            unreachable!("we always keep the receiver open")
+                        }
+                        Err(ExecutionError::StepFailed { error, .. }) => {
+                            result = Some(Err(error));
+                        }
+                    }
+                }
+                event = receiver.recv(), if !events_done => {
+                    match event {
+                        Some(event) => {
+                            match event.into_generic() {
+                                Ok(event) => {
+                                    self.payload_sender.send(
```

Comment:
Is the `payload_sender` a channel shared among all nested engines, such that events come out in one receiver?

Reply:
I think "yes and no" - the nested engine created in this function has its own payload sender (our local `sender` variable) that is distinct from `self.payload_sender`; however, any events sent to the nested `sender` are here forwarded out through `self.payload_sender` as `StepContextPayload::Nested(_)`. That would apply recursively if this nested engine itself had steps that created a nested engine: that innermost engine would have its own `sender`, which would get forwarded out via the middle `sender`, and then forwarded here through `self.payload_sender` (at this point as the moral equivalent of a `StepContextPayload::Nested(StepContextPayload::Nested(event))`, although I think it's not actually that exact type.